### PR TITLE
Feature/correct instruction documentation

### PR DESCRIPTION
**Pull request by martin-cs**
See https://github.com/diffblue/cbmc/pull/2031 for the discussion that started this.

### CODE REVIEW

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -134,17 +140,21 @@ class goto_programt
   /// - ATOMIC_BEGIN, ATOMIC_END:
   ///     When a thread executes ATOMIC_BEGIN, no thread other will be able to
   ///     execute any instruction until the same thread executes ATOMIC_END.
+  ///     Concurrency is not support by all analysis tools.
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: Nit pick: s/support/supported/

File: src/goto-programs/goto_program.h [hunk 2]
```diff
@@ -134,17 +140,21 @@ class goto_programt
...
   /// - END_FUNCTION:
-  ///     Can only occur as the last instruction of the list.
+  ///     Must occur as the last instruction of the list and nowhere else.
   /// - START_THREAD:
   ///     Create a new thread and run the code of this function starting from
   ///     targets[0]. Quite often the instruction pointed by targets[0] will be
   ///     just a FUNCTION_CALL, followed by an END_THREAD.
+  ///     Concurrency is not support by all analysis tools.
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: As above.

File: src/goto-programs/goto_program.h [hunk 3]
```diff
@@ -134,17 +140,21 @@ class goto_programt
...
   /// - END_THREAD:
   ///     Terminate the calling thread.
+  ///     Concurrency is not support by all analysis tools.
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: As above.

File: src/goto-programs/goto_program.h [hunk 4]
```diff
@@ -106,14 +106,19 @@ class goto_programt
...
+  ///     If there are multiple targets, pick one non-deterministically.
   /// - RETURN:
   ///     Set the value returned by `code` (which shall be either nil or an
   ///     instance of code_returnt) and then jump to the end of the function.
+  ///     Many analysis tools remove these instructions before they start.
   /// - DECL:
   ///     Introduces a symbol denoted by the field `code` (an instance of
   ///     code_declt), the life-time of which is bounded by a corresponding DEAD
-  ///     instruction.
+  ///     instruction.  Symbols must be DECL'd before they are used.
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: I'd refine this to "Non-static symbols must be ..."

**Review by tautschnig — APPROVED** [review]

File: src/goto-programs/goto_program.h [hunk 5]
```diff
@@ -106,14 +106,19 @@ class goto_programt
   /// respective fields in this class:
   ///
   /// - GOTO:
-  ///     if `guard` then goto `targets`
+  ///     goto `targets` if and only if `guard` is true
+  ///     If there are multiple targets, pick one non-deterministically.
```
Comment (peterschrammel) [inline src/goto-programs/goto_program.h]: Maybe add as a comment: Multiple targets are deprecated.

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
All changes applied.

**Pull request merged**